src/data/polynomial/laurent.lean

```diff
@@ -0,0 +1,186 @@
+Authors: Damiano Testa
+-/
+
+import data.polynomial.algebra_map
+
+/-!  # Laurent polynomials
+
+We introduce Laurent polynomials over a semiring `R`.  Mathematically, they are expressions of the
+form
+```latex
+\sum_{i \in \mathbb{Z}} a_i T ^ i
+```
+where the sum extends over a finite subset of `ℤ`.  Thus, negative exponents are allowed.  The
+coefficients come from the semiring `R` and the variable `T` commutes with everything.
+
+Since we are going to convert back and forth between polynomials and Laurent polynomials, we
+decided to maintain some distinction by using the symbol `T`, rather than `X`, as the variable for
+Laurent polynomials
+
+## Notation
+The symbol `R[T;T⁻¹]` stands for `laurent_polynomial R`.  We also define
+
+* `C : R →+* R[T;T⁻¹]` is the inclusion of constant polynomials, analogous to the one for `R[X]`;
+* `T : ℤ → R[T;T⁻¹]` is the sequence of powers of the variable `T`.
+
+## Implementation notes
+
+We define Laurent polynomials as `add_monoid_algebra R ℤ`.
+Thus, they are essentially `finsupp`s `ℤ →₀ R`.
+This choice differs from the current irreducible design of `polynomial`, that instead shields away
+the implementation via `finsupp`s.  It is closer to the original definition of polynomials.
+
+As a consequence, `laurent_polynomials` play well with polynomials, but there is a little roughness
+in establishing the API, since the `finsupp` implementation of `polynomial R` is well-shielded.
+
+Unlike the case of polynomials, I felt that the exponent notation was not too easy to use, as only
+natural exponents would be allowed.  Moreover, in the end, it seems likely that we should aim to
+performing computations on exponents in `ℤ` anyway and separating this via the symbol `T` seems
+convenient.
+
+I made some *heavy* use of `simp` lemmas, aiming to bring Laurent polynomials to the form
+`C a * T n`.  Any comments or suggestions for improvements is greatly appreciated!
+
+##  Future work
+Lots is missing!  I would certainly like to show that `R[T;T⁻¹]` is the localization of `R[X]`
+inverting `X`.  This should be mostly in place, given `exists_T_pow` (which is part of PR #13415).
+-/
+
+open_locale polynomial big_operators
+open polynomial add_monoid_algebra finsupp
+noncomputable theory
+
+variables {R N Z : Type*}
+
+/--  The semiring of Laurent polynomials with coefficients in the semiring `R`.
+We denote it by `R[T;T⁻¹]`.
+The ring homomorphism `C : R →+* R[T;T⁻¹]` includes `R` as the constant polynomials. -/
+abbreviation laurent_polynomial (R : Type*) [semiring R] := add_monoid_algebra R ℤ
+
+local notation R`[T;T⁻¹]`:9000 := laurent_polynomial R
+
+/--  The ring homomorphism, taking a polynomial with coefficients in `R` to a Laurent polynomial
+with coefficients in `R`. -/
+def polynomial.to_laurent [semiring R] :
+  R[X] →+* R[T;T⁻¹] :=
+begin
+  refine ring_hom.comp _ (to_finsupp_iso R).to_ring_hom,
+  exact (map_domain_ring_hom R (nat.cast_add_monoid_hom ℤ)),
```

Comment:
Definitionally this is better as
```suggestion
  exact (map_domain_ring_hom R int.of_nat_hom),
```

Reply:
Thanks for this: I am still not really sure which ones of these maps are the "correct" ones to use.